### ISSUE TITLE
GMM not converging

### ISSUE THREAD

**Issue by Bawaw** [rusty1s/pytorch_geometric#86]
Hi Matthias, 

I love how the library keeps growing, great work! I am however stuck at trying to create a simple Gaussian Mixture Model for 3d shape classification based on the MoNet paper but the loss is not decreasing. Am I doing something wrong? Should I use SHOT descriptors instead? I'm using spherical pseudo coordinates and graclus as clustering. I'll drop the code below. 

Kind regards, 
Balder 

    class MonetBaseline(torch.nn.Module):
        def __init__(self):
            super(MonetBaseline,self).__init__()
            self.conv1 = GMMConv(in_channels = 1, out_channels = 32, dim=3)
            self.batchnorm_1 = torch.nn.BatchNorm1d(32)
            self.conv2 = GMMConv(32, 64, dim=3)
            self.batchnorm_2 = torch.nn.BatchNorm1d(64)
            self.conv3 = GMMConv(64, 64, dim=3)
            self.batchnorm_3 = torch.nn.BatchNorm1d(64)
            self.fc1 = torch.nn.Linear(64, 8)
            self.fc2 = torch.nn.Linear(8, 2)

        def forward(self, data):
            data.x = F.elu(self.batchnorm_1(self.conv1(data.x, data.edge_index, data.edge_attr)))
            data.x = F.elu(self.batchnorm_2(self.conv2(data.x, data.edge_index, data.edge_attr)))
            cluster = graclus(data.edge_index, weight = data.edge_attr)
            data = max_pool(cluster, data)

            data.x = F.elu(self.batchnorm_3(self.conv3(data.x, data.edge_index, data.edge_attr)))
            cluster = graclus(data.edge_index, weight = data.edge_attr)
            x, batch = max_pool_x(cluster, data.x, data.batch)
            x = global_mean_pool(x, batch)
            x = F.elu(self.fc1(x))
            x = F.dropout(x, training=self.training) 
            x = F.elu(self.fc2(x))
            return x

**Comment by rusty1s** [comment]
Hi and sorry for answering late,

I noticed a small bug in your code, which may cause the problem:

```python
data = max_pool(cluster, data)
```
This method is a bit hard to get. I tried my best to describe it [here](https://rusty1s.github.io/pytorch_geometric/build/html/modules/nn.html#torch_geometric.nn.pool.max_pool).
However, edge features of collapsing edges will get summed up! That does not make any sense when working with 3d shapes. Instead, it is best to delete the edge features before calling max_pool, and to generate new edge features based on the newly defined positions:

```python
data.edge_attr = None
data = max_pool(cluster, data, transform=Spherical())
```

In addition, I noticed that you use the multi-dimensional edge features as weights in the [graclus](https://rusty1s.github.io/pytorch_geometric/build/html/modules/nn.html#torch_geometric.nn.pool.graclus) call:
```python
cluster = graclus(data.edge_index, weight = data.edge_attr)
```
This is not allowed and I will add an assertion to capture such mistakes. Here, `weight` is used to find the best unmarked neighbor with maximum edge weight. For your case, it is best to simply omit the weight parameter.
```python
cluster = graclus(data.edge_index)
```

**Comment by Bawaw** [comment]
Hi Matthias,

Thank you for the elaborate response! Sorry for my late reply I was without GPU for a week. Solving the bug indeed seems to make the network more robust. However, when working with a subset of 100 shapes from my dataset it still takes up to 500 epochs to converge using Adam optimization. Is this expected behavior? Using SplineCNN I could get the network to converge in well under 30 epochs. I've tried increasing the learning rate and tweaking the batch size (higher seems to work better). I'm using CrossEntropyLoss and Adam optimisation with a learning rate of 0.001 and a weight decay of 5e-4. My node features are all 1 since I'm trying to learn from shape only, I'll paste the code below. 
   
    pre_transform = T.Compose(
            [transforms.Scale(transform_scale), transforms.UniformSimilarSample(7160, sample_size), T.KNNGraph(k),
             T.Spherical(cat=False), transforms.IdentityTransform()])

    class MonetBaseline(torch.nn.Module):
        def __init__(self):
            super(MonetBaseline,self).__init__()
            self.conv1 = GMMConv(in_channels = 1, out_channels = 32, dim=3)
            self.batchnorm_1 = torch.nn.BatchNorm1d(32)
            self.conv2 = GMMConv(32, 64, dim=3)
            self.batchnorm_2 = torch.nn.BatchNorm1d(64)
            self.conv3 = GMMConv(64, 64, dim=3)
            self.batchnorm_3 = torch.nn.BatchNorm1d(64)
            self.fc1 = torch.nn.Linear(64, 8)
            self.fc2 = torch.nn.Linear(8, 2)

        def forward(self, data):
            data.x = F.elu(self.batchnorm_1(self.conv1(data.x, data.edge_index, data.edge_attr)))
            data.x = F.elu(self.batchnorm_2(self.conv2(data.x, data.edge_index, data.edge_attr)))

            cluster = graclus(data.edge_index)
            data.edge_attr = None
            data = max_pool(cluster, data, transform=T.Spherical())

            data.x = F.elu(self.batchnorm_3(self.conv3(data.x, data.edge_index, data.edge_attr)))
            cluster = graclus(data.edge_index)
            x, batch = max_pool_x(cluster, data.x, data.batch)
            x = global_mean_pool(x, batch)

            x = F.elu(self.fc1(x))
            x = F.dropout(x, training=self.training)
            x = F.elu(self.fc2(x))
            return x

Kind regards, 
Balder

**Comment by rusty1s** [comment]
I am not sure if there are numerical instabilities with GMMConv, but from what I experienced myself slower convergence seems to be natural behaviour when working with GMMConv, e.g. the authors themselves train all models in their paper for a large number of epochs.

**Comment by Bawaw** [comment]
Hi,

Sorry for opening this issue again but I really can't get this operator to converge on the simplest of problems. I've created a toy example where the goal is to distinguish a square from a parallelogram but even here the loss mainly plateau's at 0.6931. I find that sometimes I get better results if I disable the normalization in GMMConv but then as a result output can explode. Any ideas on this? I'll post the dummy example below if you're interested.

Kind regards,
Balder

    import torch
    import torch.nn.functional as F
    import torch_geometric.transforms as T
    from torch_geometric.data import DataLoader, Data
    from torch_geometric.nn import GMMConv, global_mean_pool, SplineConv

    pseudo = T.Spherical()

    edge_index1 = torch.tensor([[0,1,0,2,3,1,3,2],
                           [1,0,2,0,1,3,2,3]])

    edge_index3 = torch.tensor([[0,1,0,2,1,2],
                            [1,0,2,0,2,1]])

    x = torch.ones((4,1))
    pos_1 = torch.tensor([[50,100,10],[100,100,10],[50,50,10],[100,50,10]], dtype = torch.float32)
    pos_2 = torch.tensor([[70,100,10],[120,100,10],[50,50,10],[100,50,10]], dtype = torch.float32)
    pos_3 = torch.tensor([[50,50,10],[75,100,10],[100,50,10]], dtype = torch.float32)
    labels = torch.tensor([[0],[1]])

    # example square vs parallelogram
    data = [Data(x = x, edge_index = edge_index1, pos = pos_1, y = labels[0]), Data(x = x, edge_index =         edge_index1, pos = pos_2, y = labels[1])]
    # example square vs triangle
    #data = [Data(x = x, edge_index = edge_index1, pos = pos_1, y = labels[0]), Data(x = x, edge_index =     edge_index3, pos = pos_3, y = labels[1])]
    data = list(map(pseudo, data))

    train_loader = DataLoader(data, batch_size =2, shuffle = False)

    class GMM(torch.nn.Module):
       def __init__(self):
           super(GMM,self).__init__()
           self.conv1 = GMMConv(in_channels = 1, out_channels = 16, dim=3)
           #self.conv1 = SplineConv(1,16, dim=3, kernel_size=5)
           self.fc1 = torch.nn.Linear(16, 2)

       def forward(self, data):
           data.x = F.elu(self.conv1(data.x, data.edge_index, data.edge_attr))
           x = global_mean_pool(data.x, data.batch)
           x = self.fc1(x)
           return F.log_softmax(x, dim = 1)

    epochs = range(30000)
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    model = GMM().to(device)
    optimizer = torch.optim.Adam(model.parameters(), lr=0.0001)

    def train(epoch):
        model.train()
        epoch_loss = 0

    for data in train_loader:
        data = data.to(device)
        optimizer.zero_grad()
        out = model(data)

        _, preds = torch.max(out, 1)
        loss = F.nll_loss(out, data.y)
        epoch_loss += loss.item()
        loss.backward()
        optimizer.step()

        #print(out)
        print(preds)
        print(loss)

    for epoch in epochs:
        train(epoch)
        print('Epoch: ' + str(epoch))

**Comment by rusty1s** [comment]
Thank you so much for your effort. I revisited the `GMMConv` and might fixed slight numerical instabilities with `torch.exp`. Your toy example does now converge.
